### ISSUE TITLE
Guard condition rejects a term in coqide but not in coqc

### ISSUE THREAD

**Issue by maximedenes** [rocq-prover/rocq#9060]
Steps to reproduce:
~~~coq
Fixpoint plus (n m : nat) : nat.
Proof.
  destruct n.
  - exact m.
  - exact (S (plus n m)).
Qed. 
(* with -async-proofs on:
Recursive definition of plus is ill-formed.
In environment ...
Recursive call to plus has principal argument equal to 
"m" instead of a subterm of "m".
Recursive definition is: ... *)
~~~
- open in coqide
- jump to the end of the file
- notice coqide reports an error

**Comment by maximedenes** [comment]
Ah ah ah! The `struct` argument is not the same in the proof context and after the `Qed`... :)

**Comment by maximedenes** [comment]
In fact, I think I don't understand the intended design here. `Fixpoint` is accepted without a body *and* without a `struct` argument. But of course, at some point, the proof term will have to be refined to a `fix`, whose `struct` will be specified and maybe incorrect. There seems to be a hack to "adjust" it after the proof, but this seems broken by design, since (safe) tactics are not supposed to refine the proof term to an ill-typed term.

I suggest to make `struct` mandatory when no body is provided, and remove the corresponding hacks.

@ppedrot any thoughts?

**Comment by ppedrot** [comment]
> I suggest to make struct mandatory when no body is provided, and remove the corresponding hacks.

If you manage to do that without breaking too much backwards compatibility, I'm in favour of it.

**Comment by maximedenes** [comment]
@ppedrot in fact, this whole black magic seems to be necessary for `Lemma ... with ...` which does not even have a syntax to specify on what argument the induction is done, if I'm reading the manual correctly.

I don't understand the rationale behind not specifying what the induction is done over, which is standard even in text proofs, while the whole approach means that tactics build ill-typed terms that are patched after the fact...

**Comment by ejgallego** [comment]
You are not very precise as to whether the problem is, are you talking about `start_proof_with_initialization` or some other stuff? What is the exact codepath, are you talking instead about `do_guard` in save?

**Comment by ejgallego** [comment]
If the last, does that mean that the terminator information was lost by the STM?

**Comment by maximedenes** [comment]
I think there are two distinct problems:
- the design of `Lemma ... with` and `Fixpoint` without body not `struct` leads to tactics generating ill-typed terms (e.g. `Guarded` will fail), and let the system make arbitrary choices that change the proof context (the induction hypotheses are not the same). This is independent from the STM.
- in order to implement this design, in `Lemmas`, `adjust_guardness_conditions` tries to make the term well-typed at the last minute (before feeding the kernel) by randomly changing `struct` arguments. The STM fails to reproduce that, I guess indeed because it does not sends terminators to the slaves. I believe terminators are applied, but in master, and that's after the slave tries to check that the term is well typed. I haven't confirmed the exact codepath yet.

Bottom-line: IMHO the whole design is unsound, and break the STM expectations. Of course, it would be good to fix the two problems, by requiring the user to specify on what they want the induction to be done and making it possible to send terminators to workers (of course, they'd need to be made serializable).

**Comment by ejgallego** [comment]
I think you are spot on point 2, however note that the problem is not just confined to guardness conditions; plenty of other stuff will break due to missing "terminators". I tried to reify them recently; but I didn't arrive to something I am yet happy with.

It would be indeed good to fix point 1, but in general the system assumes that we ought to have an extensible, non-serializable platform to store arbitrary things at proof start for their comsumption at proof end and that is indeed a critical platform problem.

Personally, I think we should forbid `Marshall` and require every client to implement a proper serialization interface, for a start. This may be verbose at first [specially without deriving] but it will pay good rewards in the end.

**Comment by ejgallego** [comment]
You could try to adapt a patch I have floating around, what the patch does is to make terminators an ADT, with `Custom` and `Standard`. This doesn't work yet due to hooks, but maybe I could get lucky.

Then the idea is as follows, you can only delegate proofs if the terminator is standard.

**Comment by maximedenes** [comment]
> Then the idea is as follows, you can only delegate proofs if the terminator is standard.

Ah! That's a nice idea, to make the change incremental, because it's probably a lot of work (and it makes it possible to keep the whole thing extensible).

**Comment by ejgallego** [comment]
Ok I'll try to get that patch back in shape [not today sadly] that patch plus an adaptation of the STM should fix that example you showed me with Obligations for example, and many others that I guess are broken.

The question of allowing the extensible storage remains tho, but at least we can force the clients to mark it and degrade accordingly.

**Comment by maximedenes** [comment]
Yep, sounds very reasonable. This was urgent only because I wanted to have a CI run on stdlib with `-async-proofs on`, but we can make the stdlib scripts more robust by simply specifying the few missing `struct` arguments (3 or 4 occurrences in the whole lib). It will leave us more time to work on solving the terminator problem.

**Comment by ejgallego** [comment]
Do the proofs using obligations don't fail too?

**Comment by maximedenes** [comment]
No, they don't. But maybe they are not delegated? Or they are, but their terminators is not critical for getting well-typed terms, I haven't checked. The issue reported here is the only problem which shows up when compiling stdlib with `-async-proofs on`, and applying #9072 (otherwise the flag has no effect :D )

**Comment by ejgallego** [comment]
By the way it would be good if we could make progress on #8686 as to avoid conflicts; that PR is ready and IMHO pretty useful as to further refactor code together with #8690 (which is not ready yet)

**Comment by maximedenes** [comment]
Note that I was a bit wrong about the design. `Lemma ... with` seems to be documented as unsafe (the guard being checked only after the proof). In particular, the induction hypotheses don't change, they typically let you prove anything (it works like `fix`). What is more puzzling is that the manual advocates the use of `Guarded`, but does not mention failures can be spurious. I'll have to check if they can be, like for the `Fixpoint` example.

**Comment by maximedenes** [comment]
I think the label `part: STM` reflects only part of reality. It should be also tagged as a "proof handling" issue or something like that. What's the canonical label for such issues?

**Comment by ejgallego** [comment]
Even removing terminators doesn't fix this, thus it is indeed an STM problem w.r.t. proof state.

**Comment by gares** [comment]
I can' reproduce it V8.10+alpha-2463-g4b8a5023a1

**Comment by SkySkimmer** [comment]
Blame a984c960e63b6b1a0774cbbba63ea398100c0c3d

**Comment by gares** [comment]
ah! thanks.

**Comment by SkySkimmer** [comment]
Seems to work now
